cli_tools/gce_onestep_image_import/onestep_importer/aws_importer.go

```diff
@@ -0,0 +1,350 @@
+
+		if taskResp.ExportImageTasks[0].Status != "active" {
+			if taskResp.ExportImageTasks[0].Status != "completed" {
+				return fmt.Errorf("AWS export task wasn't completed successfully")
+			}
+			log.Println("AWS export task is completed!")
+			break
+		}
+		time.Sleep(time.Millisecond * 3000)
+	}
+
+	// 4. set exported file data
+	importer.args.exportKey = fmt.Sprintf("%v%v.vmdk", importer.args.exportFolder, tid)
+	importer.args.exportedAMIPath = fmt.Sprintf("s3://%v/%v", importer.args.exportBucket, importer.args.exportKey)
+	log.Println(fmt.Sprintf("Exported location is %v.", importer.args.exportedAMIPath))
+
+	return nil
+}
+
+// copyToGCS copies vmdk file from S3 to GCS
+func (importer *awsImporter) copyFromS3ToGCS() (string, error) {
+	// 1. get GCS path as copy destination.
+	gcsFilePath := pathutils.JoinURL(importer.args.gcsScratchBucket,
+		fmt.Sprintf("onestep-image-import-aws-%v.vmdk", pathutils.RandString(5)))
+
+	log.Println(fmt.Sprintf("Copying %v to %v.",
+		importer.args.exportedAMIPath, gcsFilePath))
+
+	// 2. create a new folder for local buffer
+	path := filepath.Join(filepath.Dir(importer.args.executablePath), fmt.Sprint("upload", pathutils.RandString(5)))
+
+	err := os.Mkdir(path, 0755)
+	if err != nil {
+		return "", err
+	}
+	defer os.RemoveAll(path)
+
+	// 3. get writer
+	bs, err := humanize.ParseBytes(uploadBufSize)
+	if err != nil {
+		return "", err
+	}
+	bkt, obj, err := storageutils.GetGCSObjectPathElements(gcsFilePath)
+	if err != nil {
+		return "", err
+	}
+	workers := int64(runtime.NumCPU())
+	writer := storageutils.NewBufferedWriter(importer.ctx, int64(bs), workers, gcsClient, importer.oauth, path, bkt, obj)
+
+	return gcsFilePath, importer.transferFile(writer)
+}
+
+// transferFile downloads S3 file and uploads to GCS concurrently
+func (importer *awsImporter) transferFile(writer io.WriteCloser) error {
+	totalSize := humanize.IBytes(uint64(importer.args.exportFileSize))
+
+	// allow max 3 chunks waiting to be uploaded
+	uploadSem := make(chan *io.ReadCloser, downloadBufNum)
+	// used to synchronize upload chunks
+	var uploadMutex sync.Mutex
+
+	// create S3 client
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+	s3Client := s3.New(sess)
+
+	// setup download
+	output, err := humanize.ParseBytes(downloadBufSize)
+	if err != nil {
+		return err
+	}
+	readSize := int64(output)
+	// take ceiling
+	readers := (importer.args.exportFileSize-1)/readSize + 1
+
+	wg := new(sync.WaitGroup)
+
+	for i := int64(0); i < readers; i++ {
+		startRange := i * readSize
+		endRange := startRange + readSize - 1
+		res, err := s3Client.GetObject(&s3.GetObjectInput{
+			Bucket: aws.String(importer.args.exportBucket),
+			Key:    aws.String(importer.args.exportKey),
+			Range:  aws.String(fmt.Sprintf("bytes=%v-%v", startRange, endRange)),
+		})
+		if err != nil {
+			log.Println(fmt.Sprintf("Error in downloading from bucket %v, key %v: %v",
```

Comment:
when err happened (and exhausted all retries), we should properly handle all resources:
1. Stop uploading. (close the writer? anything else?)
2. Cleanup uploadsem channel.
3. unblock wg.wait

Reply:
Otherwise, the mian process will stuck for a while unnecessarily

Reply:
done